packages/rocketchat-e2e/client/rocketchat.e2e.js

```diff
@@ -0,0 +1,265 @@
+/* eslint-disable no-undef */
+
+function ab2str(buf) {
+	return RocketChat.signalUtils.toString(buf);
+}
+
+function str2ab(str) {
+	return RocketChat.signalUtils.toArrayBuffer(str);
+}
+
+class E2E {
+	constructor() {
+		this.enabled = new ReactiveVar(false);
+		this.instancesByRoomId = {};
+		this.ready = new ReactiveVar(false);
+	}
+
+	isEnabled() {
+		return this.enabled.get();
+	}
+
+	isReady() {
+		return this.ready.get();
+	}
+
+	getInstanceByRoomId(roomId) {
+		if (!this.enabled.get()) {
+			return;
+		}
+
+		if (this.instancesByRoomId[roomId]) {
+			return this.instancesByRoomId[roomId];
+		}
+
+		const subscription = RocketChat.models.Subscriptions.findOne({
+			rid: roomId
+		});
+
+		if (!subscription || (subscription.t !== 'd' && subscription.t !== 'p')) {
+			return;
+		}
+
+		this.instancesByRoomId[roomId] = new RocketChat.E2E.Room(Meteor.userId(), roomId, subscription.t);
+		return this.instancesByRoomId[roomId];
+	}
+
+	startClient() {
+		// To reset all keys (local and on server):
+		// Meteor.call("emptyKeychain");
+		// RocketChat.E2EStorage.store = {}
+		// localStorage.clear();
+
+		// This is a new device for signal encryption
+		const userpass = window.prompt('Enter E2E Password');
+		console.log(userpass);
+
+		const iterations = 1000;
+		const hash = 'SHA-256';
+
+		// First, create a PBKDF2 "key" containing the password
+		window.crypto.subtle.importKey('raw', str2ab(userpass), {'name': 'PBKDF2'}, false, ['deriveKey']).
+
+		// Derive a key from the password
+			then(function(baseKey) {
+				return window.crypto.subtle.deriveKey(
+					{
+						'name': 'PBKDF2',
+						'salt': str2ab(Meteor.userId()),
+						iterations,
+						hash
+					},
+					baseKey,
+					{'name': 'AES-CBC', 'length': 256}, // Key we want
+					true, // Extractable
+					['encrypt', 'decrypt'] // For new key
+				);
+			}).
+			// Export it so we can display it
+			then(function(masterKey) {
+				RocketChat.E2EStorage.put('E2E-MasterKey', masterKey);
+				window.crypto.subtle.exportKey('jwk', masterKey).then(function(result) {
+					localStorage.setItem('E2E-MasterKey', JSON.stringify(result));
+					console.log(JSON.stringify(result));
+				});
+
+				const RSAKeys = new Promise((resolve) => {
+					Meteor.call('fetchMyKeys', function(error, result) {
+						if (result !== null && result !== undefined && EJSON.parse(result)['RSA-EPrivKey']!=='') {
+							console.log('Key found');
+							console.log(EJSON.parse(result));
+							const pubkey = EJSON.parse(result)['RSA-PubKey'];
+							const eprivkey = EJSON.parse(result)['RSA-EPrivKey'];
```

Comment:
The encrypted private key should not be stored on the server.

Reply:
Private keys are encrypted on the client and then stored in the server. This has been done to ensure that users can use any device (and more importantly, any new browser tab) to access their previous E2E encrypted message history. Otherwise, each device or browser tab would create a new set of keys and if you cleared your cache, for example, you'd lose all the messages you sent/received using that browser (as the keys would be lost).

In order to make this system secure against offline attacks including bruteforcing by a malicious server admin or attacker who gains access to the server, certain measures will be incorporated such as strong E2E password requirements, using a intensive key derivation function like bcrypt, etc.

Reply:
see [issues/9999#issuecomment-381298074](../issues/9999#issuecomment-381298074) in the #9999.